python3Packages.discordpy: refactor

GetPsyched opened the pull request:
## Description of changes

- Make consistent with newer Python package and nixpkgs rules
- Move general postPatch to lib.optionalString withVoice
- Align imports in pythonImportsCheck with upstream packages

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

GetPsyched committed:
python3Packages.discordpy: add getpsyched as maintainer

GetPsyched commented on pkgs/development/python-modules/discordpy/default.nix:
```diff
@@ -6,62 +6,62 @@
+    substituteInPlace "discord/opus.py" \
+      --replace-fail "ctypes.util.find_library('opus')" "'${libopus}/lib/libopus${stdenv.hostPlatform.extensions.sharedLibrary}'"
+
+    substituteInPlace "discord/player.py" \
+      --replace-fail "executable: str = 'ffmpeg'" "executable: str = '${lib.getExe ffmpeg}'"
+  '';
```
Loosely related, do we even need this? Can the lib not auto detect the `ffmpeg` binary without doing this replace? Also note that this replace was broken since [`discord.py` version `2.0.0`](https://github.com/Rapptz/discord.py/releases/tag/v2.0.0)

pbsds replied:
it'll work if ffmpeg is available in PATH. but propagating things that go in `bin` is not common

GetPsyched replied:
So, as long as we have `ffmpeg` as a runtime dependency, we don't need to do this substitution, right?

pbsds replied:
I tested a `nix-shell`, and ffmpeg is indeed added to PATH. go ahead, but please move `ffmpeg` from `dependencies` and over to  `propagatedBuildInputs`, since it should not be added to the python `passthru.dependencies`

GetPsyched replied:
The same should be done for `libopus`?

On a side note, I've asked upstream why they need `libopus` when `python3.opuslib` exists.

pbsds replied:
I'm done thinking out loud now, I think i still prefer patching over propagation

GetPsyched replied:
Ah, since we're adding `ffmpeg` and `libopus` in the patch anyway, we don't really need them in `dependencies` _or_ `propagatedBuildInputs`, right?

pbsds replied:
right

GetPsyched committed:
python3Packages.discord.py: refactor

- Make consistent with newer Python package and nixpkgs rules
- Move general postPatch to lib.optionalString withVoice
- Align imports in pythonImportsCheck with upstream packages

pbsds commented:
@ofborg build python311Packages.discordpy python312Packages.discordpy
